### ISSUE TITLE
flow 0.30.0 is broken with draft-js 0.8.1 (4 errors)

### ISSUE THREAD

**Issue by etrepum** [facebook/draft-js#603]
Minimal repro at https://github.com/etrepum/repro-draft-js-flow-issue with tests failing at https://travis-ci.org/etrepum/repro-draft-js-flow-issue

```
Launching Flow server for /home/travis/build/etrepum/repro-draft-js-flow-issue
Spawned flow server (pid=2802)
Logs will go to /tmp/flow/zShomezStraviszSbuildzSetrepumzSrepro-draft-js-flow-issue.log
node_modules/draft-js/lib/DraftBlockRenderConfig.js.flow:15
 15: import type React from './react';
                            ^^^^^^^^^ ./react. Required module not found

node_modules/draft-js/lib/DraftEditorProps.js.flow:47
 47:   ; blockRendererFn?(block: ContentBlock): ?Object
                         ^ Unexpected token (

node_modules/draft-js/lib/getUpdatedSelectionState.js.flow:24
 24:   if (__DEV__) {
           ^^^^^^^ identifier `__DEV__`. Could not resolve name

node_modules/fbjs/lib/Deferred.js.flow:60
 60:     Promise.prototype.done.apply(this._promise, arguments);
                           ^^^^ property `done`. Property not found in
500: declare class Promise<+R> {
     ^ Promise. See lib: /tmp/flow/flowlib_3e812a39/core.js:500


Found 4 errors
npm ERR! Test failed.  See above for more details.
```

I was able to work around this with draft-js 0.7.0 with the following flowconfig ignores (these no longer seem sufficient and were isolated to fbjs rather than draft-js, so it seems like a regression):

```
[ignore]
<PROJECT_ROOT>/node_modules/fbjs/lib/Deferred.*
<PROJECT_ROOT>/node_modules/fbjs/lib/equalsSet.*
<PROJECT_ROOT>/node_modules/fbjs/lib/everySet.*
<PROJECT_ROOT>/node_modules/fbjs/lib/PromiseMap.*
<PROJECT_ROOT>/node_modules/fbjs/lib/someSet.*
<PROJECT_ROOT>/node_modules/fbjs/lib/UnicodeBidi.*
```

**Comment by zpao** [comment]
Sorry about that & thanks for the heads up. Let's go in order:

```
node_modules/draft-js/lib/DraftBlockRenderConfig.js.flow:15
 15: import type React from './react';
                            ^^^^^^^^^ ./react. Required module not found
```

I don't think the type import should be there at all as `ReactElement` is currently globally available. I think we can fix it that way, otherwise we can ensure that the import doesn't get rewritten to include `./`
* * *
```
node_modules/draft-js/lib/DraftEditorProps.js.flow:47
 47:   ; blockRendererFn?(block: ContentBlock): ?Object
                         ^ Unexpected token (
```

Ah! Looks like babel is reprinting code incorrectly. Definitely need to fix that.
* * *
```
node_modules/draft-js/lib/getUpdatedSelectionState.js.flow:24
 24:   if (__DEV__) {
           ^^^^^^^ identifier `__DEV__`. Could not resolve name
```

We've had `__DEV__` in there a long time, not sure why it's suddenly causing issue :( As an initial workaround (which might also fix the next point in a different way), you can add this (assuming npm 3+)
```conf
  [libs]
  node_modules/fbjs/flow/lib/dev.js
```
* * *
```
node_modules/fbjs/lib/Deferred.js.flow:60
 60:     Promise.prototype.done.apply(this._promise, arguments);
                           ^^^^ property `done`. Property not found in
500: declare class Promise<+R> {
     ^ Promise. See lib: /tmp/flow/flowlib_3e812a39/core.js:500
```

You should be able to ignore `node_modules/fbjs` entirely I think. This is an unfortunate situation where fbjs knows that it should be sourcing some libs but that can't propagate up :(

**Comment by etrepum** [comment]
Ignoring all of fbjs causes a whole lot of errors (turns into a plethora of "Required module not found"), so I don't think that works around any of these problems.

I can specifically ignore Deferred.js.flow which works around `node_modules/fbjs/lib/Deferred.js.flow:60` and does't cause any other issues here. 

Adding `node_modules/fbjs/flow/lib/dev.js` to libs does work around `node_modules/draft-js/lib/getUpdatedSelectionState.js.flow:24`

Now we're just down to the two that seem unlikely to be fixable without upstream changes:

```
> repro-draft-js-flow-issue@ test /Users/bob/src/repro-draft-js-flow-issue
> flow

node_modules/draft-js/lib/DraftBlockRenderConfig.js.flow:15
 15: import type React from './react';
                            ^^^^^^^^^ ./react. Required module not found

node_modules/draft-js/lib/DraftEditorProps.js.flow:47
 47:   ; blockRendererFn?(block: ContentBlock): ?Object
                         ^ Unexpected token (


Found 2 errors
```

**Comment by zpao** [comment]
> I can specifically ignore Deferred.js.flow

I guess the other option would be to include all of `fbjs/flow/lib` in your `[libs]`, which would pull in the old `Promise` declaration. There's some risk as you would miss checks for your own misuse of Promises. We might be able to transform this away in fbjs itself, but haven't had a chance to hunt it down. Ideally we stop using non-standard methods but we own our own Promise impl so it'll be harder to get rid of those uses.


Ok, I spent more time than I really should have for the Babel transform issue, but I have https://github.com/babel/babel/pull/3663 as a potential solution.

**Comment by zpao** [comment]
I just fixed the DraftBlockRenderConfig issue fixed in #604. We're getting there…

**Comment by etrepum** [comment]
Everything looks like it works now with the latest draft and flow https://github.com/etrepum/repro-draft-js-flow-issue :+1:
